services/src/main/java/org/keycloak/protocol/oidc/par/endpoints/ParEndpoint.java

```diff
@@ -0,0 +1,397 @@
+import org.keycloak.protocol.oidc.utils.OIDCResponseType;
+import org.keycloak.protocol.oidc.utils.RedirectUtils;
+import org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequest;
+import org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequestParserProcessor;
+import org.keycloak.protocol.oidc.par.ParResponse;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.ServicesLogger;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.util.TokenUtil;
+import org.keycloak.utils.ProfileHelper;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import static org.keycloak.protocol.oidc.OIDCLoginProtocol.REQUEST_URI_PARAM;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+/**
+ * Pushed Authorization Request endpoint
+ */
+public class ParEndpoint {
+
+    public static final String PAR_CREATED_TIME = "par.created.time";
+    private static final String REQUEST_URI_TEMPLATE = "urn:ietf:params:oauth:request_uri:%s";
+
+    @Context
+    private KeycloakSession session;
+
+    @Context
+    private HttpRequest request;
+
+    @Context
+    private HttpResponse httpResponse;
+
+    @Context
+    private HttpHeaders headers;
+
+    @Context
+    private ClientConnection clientConnection;
+
+    private EventBuilder event;
+    private Cors cors;
+    private RealmModel realm;
+    private ClientModel client;
+    private AuthorizationEndpointRequest authorizationRequest;
+    private OIDCResponseType parsedResponseType;
+    private OIDCResponseMode parsedResponseMode;
+
+    // https://tools.ietf.org/html/rfc7636#section-4.2
+    private static final Pattern VALID_CODE_CHALLENGE_PATTERN = Pattern.compile("^[0-9a-zA-Z\\-\\.~_]+$");
+
+    public ParEndpoint(RealmModel realm, EventBuilder event) {
+        this.realm = realm;
+        this.event = event;
+    }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response par() {
+
+        ProfileHelper.requireFeature(Profile.Feature.PAR);
+
+        cors = Cors.add(request).auth().allowedMethods("POST").auth().exposedHeaders(Cors.ACCESS_CONTROL_ALLOW_METHODS);
+
+        event.event(EventType.PUSHED_AUTHORIZATION_REQUEST);
+
+        checkSsl();
+        checkRealm();
+        authorizeClient();
+
+        if (!OIDCAdvancedConfigWrapper.fromClientModel(client).getRequiredPushedAuthorizationRequests()) {
```

Comment:
If I understand correctly the specification chapter 6 here https://datatracker.ietf.org/doc/html/draft-ietf-oauth-par#section-6, my understanding is that:
- If `require_pushed_authorization_requests` is false, client is still able to use PAR to send request. However client can choose also the "classic" way of initiating OIDC flow with the AuthorizationEndpoint
- If `require_pushed_authorization_requests` is true, client MUST use PAR.

So if I understand correctly, it will makes sense to remove this check from this ParEndpoint as every client is allowed to use PAR if I understand correctly. But on the other hand, the AuthorizationEndpoint should have some check added, that clients with the `require_pushed_authorization_requests` set to true should be rejected from calling AuthorizationEndpoint (unless they call it with the `request_uri` parameter containing PAR URI).

WDYT?

Reply:
@mposolda Yes, I agree with your understanding. I will revise logic as you've suggested.

Reply:
Great, Thanks for the confirm